File: wlttoken/token.go
package wlttoken

import (
	"errors"
	"fmt"
	"strings"
	"time"
	"unicode"

	"github.com/KarpelesLab/apirouter"
	"github.com/KarpelesLab/base58"
	"github.com/KarpelesLab/libwallet/wltintf"
	"github.com/KarpelesLab/libwallet/wltnet"
	"github.com/KarpelesLab/outscript"
	"github.com/KarpelesLab/pobj"
	"github.com/KarpelesLab/xuid"
	"github.com/portablesql/psql"
)

// Bounds applied to untrusted token metadata (on-chain symbol / name /
// decimals, plus operator-supplied overrides). Token symbols and names
// originate from contract calls or RPC metadata an attacker controls, so
// they are sanitised and capped to defeat display-spoofing (control
// chars, bidi overrides, homographs, unbounded length) and decimals are
// bounded because they feed amount scaling.
const (
	maxTokenSymbolLen = 32
	maxTokenNameLen   = 128
	maxTokenDecimals  = 36
)

// sanitizeTokenText strips control and bidirectional/zero-width
// formatting characters from untrusted token metadata and caps the
// result to maxLen runes. This prevents token-impersonation and
// terminal/display corruption from on-chain symbol/name fields
// (e.g. a U+202E RIGHT-TO-LEFT OVERRIDE that visually reverses text,
// or homograph padding hidden behind zero-width joiners).
func sanitizeTokenText(s string, maxLen int) string {
	out := make([]rune, 0, len(s))
	for _, r := range s {
		switch {
		case r == unicode.ReplacementChar:
			// dropped: invalid UTF-8 sequence
		case unicode.IsControl(r):
			// dropped: C0/C1 control characters
		case isBidiOrInvisible(r):
			// dropped: bidi overrides / zero-width / BOM
		default:
			out = append(out, r)
			if len(out) >= maxLen {
				return strings.TrimSpace(string(out))
			}
		}
	}
	return strings.TrimSpace(string(out))
}

// isBidiOrInvisible reports whether r is a bidirectional formatting
// control or an invisible/zero-width character commonly abused to spoof
// or corrupt displayed token names.
func isBidiOrInvisible(r rune) bool {
	switch {
	case r >= '\u202A' && r <= '\u202E': // LRE, RLE, PDF, LRO, RLO
		return true
	case r >= '\u2066' && r <= '\u2069': // LRI, RLI, FSI, PDI
		return true
	case r >= '\u200B' && r <= '\u200F': // ZWSP, ZWNJ, ZWJ, LRM, RLM
		return true
	case r == '\u061C': // Arabic letter mark
		return true
	case r == '\uFEFF': // zero-width no-break space / BOM
		return true
	}
	return false
}

type token struct {
	TableName psql.Name  `sql:"Token"`
	Id        *xuid.XUID `sql:",key=PRIMARY"`
	Name      string     `sql:",type=VARCHAR,size=255"`
	Symbol    string     `sql:",type=VARCHAR,size=255"`
	Address   string     `sql:",type=VARCHAR,size=255"`
	Decimals  int        `sql:",type=INT"`
	Type      string     `sql:",type=VARCHAR,size=255"` // erc20, nft, spl-token, spl-token-2022
	Network   *xuid.XUID `sql:",type=VARCHAR,size=255"`
	Logo      string     `sql:",type=TEXT"`
	Memo      string     `sql:",type=TEXT"`
	Created   time.Time  `sql:",type=DATETIME"`
	Updated   time.Time  `sql:",type=DATETIME"`
}

func init() {
	pobj.RegisterActions[token]("Token",
		&pobj.ObjectActions{
			Fetch:  pobj.Static(apiFetchToken),
			List:   pobj.Static(apiListToken),
			Create: pobj.Static(apiCreateToken),
		},
	)
}

func (t *token) validate(e wltintf.Env) error {
	if t.Network == nil {
		return errors.New("Network is required")
	}
	if t.Address == "" {
		return errors.New("Address is required")
	}

	net, err := wltnet.NetworkById(e, t.Network)
	if err != nil {
		return fmt.Errorf("invalid network: %w", err)
	}

	switch net.Type {
	case "evm":
		addr, err := outscript.ParseEvmAddress(t.Address)
		if err != nil {
			return fmt.Errorf("invalid EVM address: %w", err)
		}
		t.Address, err = addr.Address()
		if err != nil {
			return fmt.Errorf("failed to normalize EVM address: %w", err)
		}
		if t.Type == "" {
			t.Type = "erc20"
		}
	case "solana":
		decoded, err := base58.Bitcoin.Decode(t.Address)
		if err != nil {
			return fmt.Errorf("invalid Solana address: %w", err)
		}
		if len(decoded) != 32 {
			return errors.New("invalid Solana address: must be 32 bytes")
		}
		t.Address = base58.Bitcoin.Encode(decoded)
		if t.Type == "" {
			t.Type = "spl-token"
		}
	default:
		return fmt.Errorf("tokens are not supported on %s networks", net.Type)
	}

	// Sanitise display metadata — Symbol/Name may originate from
	// untrusted on-chain sources (discovery, swap EnsureToken) and are
	// otherwise persisted verbatim.
	t.Symbol = sanitizeTokenText(t.Symbol, maxTokenSymbolLen)
	t.Name = sanitizeTokenText(t.Name, maxTokenNameLen)

	if t.Decimals < 0 || t.Decimals > maxTokenDecimals {
		return fmt.Errorf("Decimals must be between 0 and %d", maxTokenDecimals)
	}

	return nil
}

func (t *token) save(e wltintf.Env) error {
	now := time.Now()
	if t.Created.IsZero() {
		t.Created = now
	}
	t.Updated = now
	return psql.Replace(e, t)
}

func TokenById(e wltintf.Env, id *xuid.XUID) (*token, error) {
	if id.Prefix != "tok" {
		return nil, fmt.Errorf("invalid key for token: %s", id.Prefix)
	}
	return wltintf.ByPrimaryKey[token](e, id)
}

// EnsureToken returns the Token row for (networkId, address), creating
// one with the supplied metadata if none exists yet. Used by callers
// that receive a token they don't necessarily already track (e.g.
// wltswap when a swap lands on a token the user has never held before)
// to surface the new asset in the user's token list without requiring
// an explicit `Token:create` round-trip.
//
// Address is normalised the same way `Token:create` normalises it
// (EVM checksum case, Solana base58 round-trip via [token.validate]),
// so callers can pass whatever shape the upstream provider returned.
//
// Empty `symbol` / `name` are accepted; UI fills in a fallback display
// from the address. `tokenType` left empty defaults to the chain's
// canonical type ("erc20" / "spl-token") in validate().
//
// NOTE: there is no `UNIQUE(Network, Address)` constraint on the Token
// table today, so a transient DB error between lookup and insert could
// produce a duplicate row. Acceptable for the swap path — the user can
// dedupe via Token:delete — but worth tightening if EnsureToken grows
// more callers.
func EnsureToken(e wltintf.Env, networkId *xuid.XUID, address, symbol, name string, decimals int, tokenType string) (*token, error) {
	if networkId == nil {
		return nil, errors.New("network is required")
	}
	if address == "" {
		return nil, errors.New("address is required")
	}
	t := &token{
		Network:  networkId,
		Address:  address,
		Symbol:   symbol,
		Name:     name,
		Decimals: decimals,
		Type:     tokenType,
	}
	if err := t.validate(e); err != nil {
		return nil, err
	}

	if existing, err := psql.Get[token](e, map[string]any{
		"Network": networkId,
		"Address": t.Address,
	}); err == nil {
		return existing, nil
	}

	var err error
	t.Id, err = xuid.NewRandom("tok")
	if err != nil {
		return nil, err
	}
	if err := t.save(e); err != nil {
		return nil, err
	}
	return t, nil
}

// Exported accessors for external callers (e.g. wlttx for ERC-20 transfer encoding).

func (t *token) GetAddress() string     { return t.Address }
func (t *token) GetDecimals() int       { return t.Decimals }
func (t *token) GetType() string        { return t.Type }
func (t *token) GetNetwork() *xuid.XUID { return t.Network }
func (t *token) GetName() string        { return t.Name }
func (t *token) GetSymbol() string      { return t.Symbol }

// LookupTokenByMint returns the Token row for (networkId, address), or
// nil if no row exists. Distinguishes "not found" (nil, nil) from
// "lookup failed" (nil, err). Used by callers (e.g. wltbase asset
// enrichment) that want to overlay user-saved metadata on top of a
// chain-fetched balance list without erroring out when the mint isn't
// tracked.
// TokensByNetwork returns every registered Token row for the given network.
// Used by Asset:list to enumerate the user's ERC-20 tokens (the Solana path
// discovers tokens on-chain instead; EVM has no cheap on-chain enumeration,
// so the registry the user builds via Token:create / Token:discoverToken is
// the source of truth).
func TokensByNetwork(e wltintf.Env, networkId *xuid.XUID) ([]*token, error) {
	if networkId == nil {
		return nil, nil
	}
	return psql.Fetch[token](e, map[string]any{"Network": networkId})
}

func LookupTokenByMint(e wltintf.Env, networkId *xuid.XUID, address string) (*token, error) {
	if networkId == nil || address == "" {
		return nil, nil
	}
	t, err := psql.Get[token](e, map[string]any{
		"Network": networkId,
		"Address": address,
	})
	if err != nil {
		return nil, nil
	}
	return t, nil
}

func (t *token) ApiDelete(ctx *apirouter.Context) error {
	e := wltintf.GetEnv(ctx)
	if e == nil {
		return errors.New("failed to get env")
	}

	_, err := psql.ForceDelete[token](e, map[string]any{"Id": t.Id})
	return err
}

func (t *token) ApiUpdate(ctx *apirouter.Context) error {
	e := wltintf.GetEnv(ctx)
	if e == nil {
		return errors.New("failed to get env")
	}

	updated := false

	if v, ok := apirouter.GetParam[string](ctx, "Name"); ok {
		t.Name = sanitizeTokenText(v, maxTokenNameLen)
		updated = true
	}
	if v, ok := apirouter.GetParam[string](ctx, "Symbol"); ok {
		t.Symbol = sanitizeTokenText(v, maxTokenSymbolLen)
		updated = true
	}
	if v, ok := apirouter.GetParam[int](ctx, "Decimals"); ok {
		if v < 0 || v > maxTokenDecimals {
			return fmt.Errorf("Decimals must be between 0 and %d", maxTokenDecimals)
		}
		t.Decimals = v
		updated = true
	}
	if v, ok := apirouter.GetParam[string](ctx, "Logo"); ok {
		t.Logo = v
		updated = true
	}
	if v, ok := apirouter.GetParam[string](ctx, "Memo"); ok {
		t.Memo = v
		updated = true
	}
	if v, ok := apirouter.GetParam[string](ctx, "Type"); ok {
		t.Type = v
		updated = true
	}

	if !updated {
		return nil
	}
	return t.save(e)
}

func apiListToken(ctx *apirouter.Context) (any, error) {
	return wltintf.ListHelper[token](ctx, "Name ASC", "Name", "Symbol", "Address", "Type")
}

func apiFetchToken(ctx *apirouter.Context, in struct{ Id string }) (any, error) {
	e := wltintf.GetEnv(ctx)
	if e == nil {
		return nil, errors.New("failed to get env")
	}

	id, err := xuid.Parse(in.Id)
	if err != nil {
		return nil, err
	}

	return TokenById(e, id)
}

// resolveNetworkRef accepts a network reference in either form the clients
// use and returns the network's xuid:
//
//   - a network xuid ("net-…"), or
//   - the canonical "<type>.<chainId>" key (e.g. "evm.137", "solana.mainnet")
//     — what Asset.network and the Dart Token API send.
//
// The canonical form is the one that previously blew up as
// "invalid UUID length: 7" (e.g. "evm.137") because Token:create /
// Token:discoverToken parsed Network directly as an xuid. Network existence
// is validated by the caller (NetworkById); this only maps ref → id.
func resolveNetworkRef(ref string) (*xuid.XUID, error) {
	ref = strings.TrimSpace(ref)
	if ref == "" {
		return nil, errors.New("Network is required")
	}
	// A real network xuid has dashes ("net-…"); the canonical key uses dots.
	if id, err := xuid.Parse(ref); err == nil {
		return id, nil
	}
	if i := strings.IndexByte(ref, '.'); i > 0 && i < len(ref)-1 {
		return wltnet.NetworkIdForTypeAndChainId(ref[:i], ref[i+1:]), nil
	}
	return nil, fmt.Errorf("invalid network reference %q (want a net-… id or \"<type>.<chainId>\")", ref)
}

// apiCreateToken takes Network as a string (xuid OR "<type>.<chainId>") rather
// than binding it straight into token.Network (*xuid.XUID), which rejected the
// canonical "evm.137" form with "invalid UUID length: 7".
func apiCreateToken(ctx *apirouter.Context, in struct {
	Name     string
	Symbol   string
	Address  string
	Decimals int
	Type     string
	Network  string
	Logo     string
	Memo     string
}) (any, error) {
	e := wltintf.GetEnv(ctx)
	if e == nil {
		return nil, errors.New("failed to get env")
	}

	netId, err := resolveNetworkRef(in.Network)
	if err != nil {
		return nil, err
	}

	t := &token{
		Name:     in.Name,
		Symbol:   in.Symbol,
		Address:  in.Address,
		Decimals: in.Decimals,
		Type:     in.Type,
		Network:  netId,
		Logo:     in.Logo,
		Memo:     in.Memo,
	}

	if err := t.validate(e); err != nil {
		return nil, err
	}

	t.Id, err = xuid.NewRandom("tok")
	if err != nil {
		return nil, err
	}

	if err := t.save(e); err != nil {
		return nil, err
	}

	return t, nil
}

File: rust/src/models/token.rs
//! wlttoken (Token model) — port of the Go `wlttoken` package.
//!
//! Read surface: fetch/list of known tokens (PascalCase JSON, plain struct).
//! Token creation discovers on-chain metadata (RPC) and is deferred (POST 501).

use serde::{Deserialize, Serialize};
use serde_json::Value;
use xuid::Xuid;

use crate::{Env, Result, SqlValue};

/// Bounds applied to untrusted token metadata (on-chain symbol / name /
/// decimals, plus operator-supplied overrides). Symbols/names originate from
/// contract calls or RPC metadata an attacker controls, so they are sanitised
/// and capped to defeat display-spoofing; decimals are bounded because they
/// feed amount scaling. Matches Go `maxTokenSymbolLen` / `maxTokenNameLen` /
/// `maxTokenDecimals`.
pub const MAX_TOKEN_SYMBOL_LEN: usize = 32;
pub const MAX_TOKEN_NAME_LEN: usize = 128;
pub const MAX_TOKEN_DECIMALS: i64 = 36;

const TABLE_DDL: &str = r#"CREATE TABLE IF NOT EXISTS "Token" ("Id" text, "Name" text, "Symbol" text, "Address" text, "Decimals" integer, "Type" text, "Network" text, "Logo" text, "Memo" text, "Created" text, "Updated" text, PRIMARY KEY ("Id"));"#;
const COLS: &str = r#""Id", "Name", "Symbol", "Address", "Decimals", "Type", "Network", "Logo", "Memo", "Created", "Updated""#;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Token {
    #[serde(rename = "Id", default)]
    pub id: String,
    #[serde(rename = "Name", default)]
    pub name: String,
    #[serde(rename = "Symbol", default)]
    pub symbol: String,
    #[serde(rename = "Address", default)]
    pub address: String,
    #[serde(rename = "Decimals", default)]
    pub decimals: i64,
    #[serde(rename = "Type", default)]
    pub kind: String, // erc20 | nft | spl-token | spl-token-2022
    #[serde(rename = "Network", default)]
    pub network: String,
    #[serde(rename = "Logo", default)]
    pub logo: String,
    #[serde(rename = "Memo", default)]
    pub memo: String,
    #[serde(rename = "Created", default)]
    pub created: String,
    #[serde(rename = "Updated", default)]
    pub updated: String,
}

pub fn init(env: &Env) -> Result<()> {
    env.ensure_table(TABLE_DDL)
}

pub fn fetch(env: &Env, id: &str) -> Result<Option<Token>> {
    let sql = format!(r#"SELECT {COLS} FROM "Token" WHERE "Id" = ?1"#);
    let rows = env.query(&sql, vec![SqlValue::Text(id.to_owned())])?;
    Ok(rows.first().map(|r| row_to_token(r)))
}

pub fn list(env: &Env) -> Result<Vec<Token>> {
    let sql = format!(r#"SELECT {COLS} FROM "Token" ORDER BY "Symbol" ASC"#);
    let rows = env.query(&sql, Vec::new())?;
    Ok(rows.iter().map(|r| row_to_token(r)).collect())
}

/// Every registered Token row for `network_id` (port of Go `TokensByNetwork`).
/// Used by `Asset:list` to enumerate the user's ERC-20 tokens: EVM has no cheap
/// on-chain owner→tokens query (unlike Solana, which discovers token accounts
/// on-chain), so the registry the user builds via `Token:create` /
/// `Token:discoverToken` (or swap EnsureToken) is the source of truth. An empty
/// id yields no rows (Go returns nil for a nil id).
pub fn tokens_by_network(env: &Env, network_id: &str) -> Result<Vec<Token>> {
    if network_id.is_empty() {
        return Ok(Vec::new());
    }
    let sql = format!(r#"SELECT {COLS} FROM "Token" WHERE "Network" = ?1"#);
    let rows = env.query(&sql, vec![SqlValue::Text(network_id.to_owned())])?;
    Ok(rows.iter().map(|r| row_to_token(r)).collect())
}

/// Find the registered token on `network_id` whose on-chain address / mint
/// matches `addr`, comparing via a base58 round-trip so equivalent encodings
/// collapse (port of Go `LookupTokenByMint`). Returns `None` when no row
/// matches. Used to resolve a canonical "<type>.<chainId>.<mint>" asset key to
/// its Token row on the SPL send path.
pub fn lookup_by_mint(env: &Env, network_id: &str, addr: &str) -> Result<Option<Token>> {
    let want = bs58::decode(addr).into_vec().ok();
    for t in tokens_by_network(env, network_id)? {
        if t.address == addr {
            return Ok(Some(t));
        }
        // Fall back to a byte-level compare so a differently-cased/padded but
        // equivalent base58 encoding still matches.
        if let (Some(a), Some(b)) = (&want, bs58::decode(&t.address).into_vec().ok()) {
            if *a == b {
                return Ok(Some(t));
            }
        }
    }
    Ok(None)
}

/// Resolve a network reference to the network's id (xuid string), accepting
/// either form the clients use (port of Go `resolveNetworkRef`):
///
///   - a network xuid ("net-…") passes through unchanged;
///   - the canonical "<type>.<chainId>" key (e.g. "evm.137", "solana.mainnet")
///     — the form `Asset.network` and the Dart Token API send — maps to the
///     deterministic network id.
///
/// The canonical form is the one that previously failed as "invalid UUID
/// length: 7" (e.g. "evm.137") because `Network` was parsed straight as an
/// xuid. Network existence is validated by the caller (`network::fetch`), not
/// here; this only maps ref → id.
pub fn resolve_network_ref(reference: &str) -> Result<String> {
    let reference = reference.trim();
    if reference.is_empty() {
        return Err(crate::Error::Env("Network is required".into()));
    }
    // A real network xuid parses under the "net" prefix; the canonical key uses
    // dots and won't.
    if Xuid::parse_prefix(reference, "net").is_ok() {
        return Ok(reference.to_owned());
    }
    if let Some((typ, chain)) = reference.split_once('.') {
        if !typ.is_empty() && !chain.is_empty() {
            return Ok(crate::models::network::network_id_for(typ, chain));
        }
    }
    Err(crate::Error::Env(format!(
        "invalid network reference {reference:?} (want a net-… id or \"<type>.<chainId>\")"
    )))
}

fn row_to_token(row: &[SqlValue]) -> Token {
    let text = |i: usize| row.get(i).and_then(|v| v.as_text()).unwrap_or("").to_owned();
    Token {
        id: text(0),
        name: text(1),
        symbol: text(2),
        address: text(3),
        decimals: row.get(4).and_then(|v| v.as_i64()).unwrap_or(0),
        kind: text(5),
        network: text(6),
        logo: text(7),
        memo: text(8),
        created: text(9),
        updated: text(10),
    }
}

/// Validate + normalize a token (port of Go `token.validate`): the network must
/// exist and support tokens, the address is normalized per chain (EVM checksum
/// case / Solana base58 round-trip), the type defaults to the chain canonical
/// ("erc20" / "spl-token"), display metadata is sanitised and decimals bounded.
pub fn validate(env: &Env, t: &mut Token) -> Result<()> {
    if t.network.is_empty() {
        return Err(crate::Error::Env("Network is required".into()));
    }
    if t.address.is_empty() {
        return Err(crate::Error::Env("Address is required".into()));
    }
    let net = crate::models::network::fetch(env, &t.network)?
        .ok_or_else(|| crate::Error::Env(format!("invalid network: {}", t.network)))?;
    match net.kind.as_str() {
        "evm" => {
            t.address = normalize_evm_address(&t.address)?;
            if t.kind.is_empty() {
                t.kind = "erc20".to_owned();
            }
        }
        "solana" => {
            t.address = normalize_solana_address(&t.address)?;
            if t.kind.is_empty() {
                t.kind = "spl-token".to_owned();
            }
        }
        other => {
            return Err(crate::Error::Env(format!(
                "tokens are not supported on {other} networks"
            )))
        }
    }
    // Sanitise display metadata — Symbol/Name may originate from untrusted
    // on-chain sources and are otherwise persisted verbatim.
    t.symbol = sanitize_token_text(&t.symbol, MAX_TOKEN_SYMBOL_LEN);
    t.name = sanitize_token_text(&t.name, MAX_TOKEN_NAME_LEN);
    if t.decimals < 0 || t.decimals > MAX_TOKEN_DECIMALS {
        return Err(crate::Error::Env(format!(
            "Decimals must be between 0 and {MAX_TOKEN_DECIMALS}"
        )));
    }
    Ok(())
}

/// Create a token (port of Go `apiCreateToken`): validate/normalize, assign a
/// random `tok` id, persist, and return the created row.
pub fn create(env: &Env, mut t: Token) -> Result<Token> {
    // Accept the canonical "<type>.<chainId>" network ref (e.g. "evm.137") the
    // Dart Token API sends, in addition to a stored net-… xuid, resolving it to
    // the network id (Go apiCreateToken / resolveNetworkRef). Storing the
    // resolved id — not "evm.137" verbatim — is also what lets Asset:list find
    // the row via tokens_by_network.
    t.network = resolve_network_ref(&t.network)?;
    validate(env, &mut t)?;
    t.id = Xuid::new_random("tok").to_string();
    save(env, &mut t)?;
    Ok(t)
}

/// Apply the mutable fields Go `token.ApiUpdate` allows (Name, Symbol,
/// Decimals, Logo, Memo, Type) from `params` and persist. Name/Symbol are
/// sanitised; Decimals is bounds-checked. A no-op update returns the row
/// unchanged.
pub fn update(env: &Env, id: &str, params: &Value) -> Result<Token> {
    let mut t = fetch(env, id)?
        .ok_or_else(|| crate::Error::Env(format!("token not found: {id}")))?;
    let mut updated = false;
    if let Some(v) = params.get("Name").and_then(Value::as_str) {
        t.name = sanitize_token_text(v, MAX_TOKEN_NAME_LEN);
        updated = true;
    }
    if let Some(v) = params.get("Symbol").and_then(Value::as_str) {
        t.symbol = sanitize_token_text(v, MAX_TOKEN_SYMBOL_LEN);
        updated = true;
    }
    if let Some(v) = params.get("Decimals").and_then(Value::as_i64) {
        if v < 0 || v > MAX_TOKEN_DECIMALS {
            return Err(crate::Error::Env(format!(
                "Decimals must be between 0 and {MAX_TOKEN_DECIMALS}"
            )));
        }
        t.decimals = v;
        updated = true;
    }
    if let Some(v) = params.get("Logo").and_then(Value::as_str) {
        t.logo = v.to_owned();
        updated = true;
    }
    if let Some(v) = params.get("Memo").and_then(Value::as_str) {
        t.memo = v.to_owned();
        updated = true;
    }
    if let Some(v) = params.get("Type").and_then(Value::as_str) {
        t.kind = v.to_owned();
        updated = true;
    }
    if !updated {
        return Ok(t);
    }
    save(env, &mut t)?;
    Ok(t)
}

/// Delete a token by id (port of Go `token.ApiDelete` — ForceDelete by Id).
pub fn delete(env: &Env, id: &str) -> Result<()> {
    env.exec(r#"DELETE FROM "Token" WHERE "Id" = ?1"#, vec![SqlValue::Text(id.to_owned())])?;
    Ok(())
}

/// Insert or replace a token row (Go `token.save` via psql.Replace). Sets
/// Created (first save) / Updated timestamps on `t`.
fn save(env: &Env, t: &mut Token) -> Result<()> {
    let now = crate::now_rfc3339();
    if t.created.is_empty() {
        t.created = now.clone();
    }
    t.updated = now;
    env.exec(r#"DELETE FROM "Token" WHERE "Id" = ?1"#, vec![SqlValue::Text(t.id.clone())])?;
    env.exec(
        &format!(r#"INSERT INTO "Token" ({COLS}) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11)"#),
        vec![
            SqlValue::Text(t.id.clone()),
            SqlValue::Text(t.name.clone()),
            SqlValue::Text(t.symbol.clone()),
            SqlValue::Text(t.address.clone()),
            SqlValue::Int(t.decimals),
            SqlValue::Text(t.kind.clone()),
            SqlValue::Text(t.network.clone()),
            SqlValue::Text(t.logo.clone()),
            SqlValue::Text(t.memo.clone()),
            SqlValue::Text(t.created.clone()),
            SqlValue::Text(t.updated.clone()),
        ],
    )?;
    Ok(())
}

/// Normalize an EVM token address to its EIP-55 checksummed form (Go uses
/// `outscript.ParseEvmAddress(..).Address()`).
fn normalize_evm_address(addr: &str) -> Result<String> {
    let hex = addr.strip_prefix("0x").or_else(|| addr.strip_prefix("0X")).unwrap_or(addr);
    if hex.len() != 40 || !hex.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err(crate::Error::Env(format!("invalid EVM address: {addr}")));
    }
    let bytes: Vec<u8> = (0..40)
        .step_by(2)
        .map(|i| u8::from_str_radix(&hex[i..i + 2], 16).unwrap())
        .collect();
    Ok(outscript::address::eip55(&bytes))
}

/// Normalize a Solana token mint address via a base58 round-trip, verifying it
/// decodes to exactly 32 bytes (Go `base58.Bitcoin.Decode` + length check).
fn normalize_solana_address(addr: &str) -> Result<String> {
    let decoded = bs58::decode(addr)
        .into_vec()
        .map_err(|e| crate::Error::Env(format!("invalid Solana address: {e}")))?;
    if decoded.len() != 32 {
        return Err(crate::Error::Env("invalid Solana address: must be 32 bytes".into()));
    }
    Ok(bs58::encode(&decoded).into_string())
}

/// Strip control / replacement / bidi-invisible characters, cap the kept runes
/// at `max`, then trim surrounding whitespace (Go `sanitizeTokenText`).
pub fn sanitize_token_text(s: &str, max: usize) -> String {
    let mut out = String::new();
    let mut kept = 0usize;
    for c in s.chars() {
        if c == '\u{FFFD}' || c.is_control() || is_bidi_or_invisible(c) {
            continue;
        }
        out.push(c);
        kept += 1;
        if kept >= max {
            break;
        }
    }
    out.trim().to_owned()
}

/// Bidi formatting controls / zero-width / BOM commonly abused to spoof token
/// names (Go `isBidiOrInvisible`).
fn is_bidi_or_invisible(c: char) -> bool {
    matches!(c,
        '\u{202A}'..='\u{202E}'   // LRE RLE PDF LRO RLO
        | '\u{2066}'..='\u{2069}' // LRI RLI FSI PDI
        | '\u{200B}'..='\u{200F}' // ZWSP ZWNJ ZWJ LRM RLM
        | '\u{061C}'              // Arabic letter mark
        | '\u{FEFF}'              // BOM / ZWNBSP
    )
}
